File: app/Models/models.dart/animal.dart
import 'package:isar/isar.dart';
import 'animal_identifier.dart';
import 'person_role.dart';
import 'event.dart';

part 'animal.g.dart';

@Collection()
class Animal {
  Id id = Isar.autoIncrement;

  int? createdBy;
  int? premisesId;
  String? species;
  String? sex;
  DateTime? birthDate;
  String? lifeStatus;
  String? uid;
  int? version;

  DateTime? createdAt;
  DateTime? updatedAt;
  DateTime? deletedAt;

  final creator = IsarLink<dynamic>();
  final premise = IsarLink<dynamic>();
  final identifiers = IsarLinks<AnimalIdentifier>();
  final personRoles = IsarLinks<PersonRole>();
  final events = IsarLinks<Event>();

  Animal({
    this.createdBy,
    this.premisesId,
    this.species,
    this.sex,
    this.birthDate,
    this.lifeStatus,
    this.uid,
    this.version,
    this.createdAt,
    this.updatedAt,
    this.deletedAt,
  });

  factory Animal.fromJson(Map<String, dynamic> json) {
    return Animal(
      createdBy: json['created_by'] as int?,
      premisesId: json['premises_id'] as int?,
      species: json['species'] as String?,
      sex: json['sex'] as String?,
      birthDate: json['birth_date'] == null
          ? null
          : DateTime.parse(json['birth_date'] as String),
      lifeStatus: json['life_status'] as String?,
      uid: json['uid'] as String?,
      version: json['version'] is int
          ? json['version'] as int
          : (json['version'] != null
              ? int.tryParse(json['version'].toString())
              : null),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'created_by': createdBy,
      'premises_id': premisesId,
      'species': species,
      'sex': sex,
      'birth_date': birthDate?.toIso8601String(),
      'life_status': lifeStatus,
      'uid': uid,
      'version': version,
      'created_at': createdAt?.toIso8601String(),
      'updated_at': updatedAt?.toIso8601String(),
      'deleted_at': deletedAt?.toIso8601String(),
    };
  }
}

File: app/Models/Animal.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\SoftDeletes;
use Illuminate\Database\Eloquent\Factories\HasFactory;

class Animal extends Model
{
    /** @use HasFactory<\Database\Factories\AnimalFactory> */
    use HasFactory;
    use SoftDeletes;

    const SPECIES = ['OVINE', 'CAPRINE'];
    const SEXES = ['M', 'F'];
    const LIFE_STATUSES = ['ALIVE', 'DEAD', 'SOLD'];

    protected $fillable = [

        'created_by',
        'premises_id',
        'species',
        'sex',
        'birth_date',
        'life_status',
        'uid',
        'version'
    ];
    public function scopeSearch($query, ?string $term)
    {
        $term = trim((string) $term);
        if ($term === '') {
            return $query;
        }
        $like = '%' . str_replace(['%', '_'], ['\\%', '\\_'], $term) . '%';

        return $query->where(function ($q) use ($like) {
            $q->where('uid', 'like', $like)
                ->orWhere('birth_date', 'like', $like)
                ->orWhereHas('identifiers', function ($query) use ($like) {
                    $query->where('code', 'like', $like);
                });
        });
    }
    public static function getTableName()
    {
        return (new self)->getTable();
    }

    public function creator()
    {
        return $this->belongsTo(User::class, 'created_by');
    }

    public function premise()
    {
        return $this->belongsTo(Premise::class, 'premises_id');
    }

    public function identifiers()
    {
        return $this->hasMany(AnimalIdentifier::class);
    }

    public function personRoles()
    {
        return $this->hasMany(PersonRole::class);
    }

    public function events()
    {
        return $this->hasMany(Event::class);
    }

    public function motherReproductionEvents()
    {
        return $this->hasMany(ReproductionEvent::class, 'mother_id');
    }

    public function fatherReproductionEvents()
    {
        return $this->hasMany(ReproductionEvent::class, 'father_id');
    }

    public function motherBirthEvents()
    {
        return $this->hasMany(BirthEvent::class, 'mother_id');
    }

    public function fatherBirthEvents()
    {
        return $this->hasMany(BirthEvent::class, 'father_id');
    }

    public function performanceRecords()
    {
        return $this->hasMany(PerformanceRecord::class);
    }
}
